crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+                    for child_index in [left, right] {
+                        let mut block = self.get_block(child_index)?;
+                        block.node.parent = Some(0);
+                        self.insert_entry_to_blob(child_index, block.to_bytes())?;
+                    }
+                }
+            };
+
+            self.free_indexes.push(sibling_index);
+
+            return Ok(());
+        };
+
+        self.free_indexes.push(parent_index);
+        let mut grandparent_block = self.get_block(grandparent_index).unwrap();
+
+        sibling_block.node.parent = Some(grandparent_index);
+        let range = self.get_block_range(sibling_index);
+        self.blob[range].copy_from_slice(&sibling_block.to_bytes());
+
+        match grandparent_block.node.specific {
+            NodeSpecific::Internal {
+                ref mut left,
+                ref mut right,
+                ..
+            } => match parent_index {
+                x if x == *left => *left = sibling_index,
+                x if x == *right => *right = sibling_index,
+                _ => panic!(),
+            },
+            NodeSpecific::Leaf { .. } => panic!(),
+        };
+        let range = self.get_block_range(grandparent_index);
+        self.blob[range].copy_from_slice(&grandparent_block.to_bytes());
+
+        self.mark_lineage_as_dirty(grandparent_index)?;
+
+        Ok(())
+    }
+
+    // fn update_parent(&mut self, index: TreeIndex, parent: Option<TreeIndex>) -> Result<(), String> {
+    //     let range = self.get_block_range(index);
+    //
+    //     let mut node = self.get_node(index)?;
+    //     node.parent = parent;
+    //     self.blob[range].copy_from_slice(&node.to_bytes());
+    //
+    //     Ok(())
+    // }
+
+    // fn update_left(&mut self, index: TreeIndex, left: Option<TreeIndex>) -> Result<(), String> {
+    //     let range = self.get_block_range(index);
+    //
+    //     let mut node = self.get_node(index)?;
+    //     node.left = left;
+    //     self.blob[range].copy_from_slice(&node.to_bytes());
+    //
+    //     Ok(())
+    // }
+
+    fn mark_lineage_as_dirty(&mut self, index: TreeIndex) -> Result<(), String> {
+        let mut next_index = Some(index);
+
+        while let Some(this_index) = next_index {
+            let mut block = Block::from_bytes(self.get_block_bytes(this_index)?, this_index)?;
+            block.metadata.dirty = true;
+            self.insert_entry_to_blob(this_index, block.to_bytes())?;
+            next_index = block.node.parent;
+        }
+
+        Ok(())
+    }
+
+    fn get_new_index(&mut self) -> TreeIndex {
+        match self.free_indexes.pop() {
+            None => {
+                self.last_allocated_index += 1;
+                self.last_allocated_index - 1
+            }
+            Some(new_index) => new_index,
+        }
+    }
+
+    fn get_random_leaf_node_from_bytes(&self, seed_bytes: Vec<u8>) -> Result<Node, String> {
```

Comment:
this looks like a test-only function. If it is, you could add a `#[cfg(test)]` to its definition

Reply:
it's called from `.insert()`

```rust
        // TODO: make this a parameter so we have one insert call where you specify the location
        let old_leaf = self.get_random_leaf_node_from_bytes(Vec::from(key_value.to_be_bytes()))?;
```

though, as per the comment, i think i want the outer layer (python datalayer code) to just first call this 'get a leaf' method and then pass it to the insert function.  then we don't need an 'insert to an automatic location' method and a separate (or parameter controlled) 'insert to this specific location' method.  just one method to auto-pick a location and one method to insert to a specific location.

Reply:
For future reference (sounds maybe not applicable here) I think all test related stuff should go inside a test module, so you don't have to repeat `#[cfg(test)]`